core/clients/key_flow.go

```diff
@@ -0,0 +1,359 @@
+	if refreshTokenIsValid {
+		return c.createAccessTokenWithRefreshToken()
+	}
+	return c.createAccessToken()
+}
+
+// createAccessToken creates an access token using self signed JWT
+func (c *KeyFlow) createAccessToken() (err error) {
+	grant := "urn:ietf:params:oauth:grant-type:jwt-bearer"
+	assertion, err := c.generateSelfSignedJWT()
+	if err != nil {
+		return err
+	}
+	res, err := c.requestToken(grant, assertion)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		tempErr := res.Body.Close()
+		if tempErr != nil {
+			err = fmt.Errorf("closing request access token response: %w", tempErr)
+		}
+	}()
+	return c.parseTokenResponse(res)
+}
+
+// createAccessTokenWithRefreshToken creates an access token using
+// an existing pre-validated refresh token
+func (c *KeyFlow) createAccessTokenWithRefreshToken() (err error) {
+	res, err := c.requestToken("refresh_token", c.token.RefreshToken)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		tempErr := res.Body.Close()
+		if tempErr != nil {
+			err = fmt.Errorf("closing request access token with refresh token response: %w", tempErr)
+		}
+	}()
+	return c.parseTokenResponse(res)
+}
+
+// generateSelfSignedJWT generates JWT token
+func (c *KeyFlow) generateSelfSignedJWT() (string, error) {
+	claims := jwt.MapClaims{
+		"iss": c.key.Credentials.Iss,
+		"sub": c.key.Credentials.Sub,
+		"jti": uuid.New(),
+		"aud": c.key.Credentials.Aud,
+		"iat": jwt.NewNumericDate(time.Now()),
+		"exp": jwt.NewNumericDate(time.Now().Add(10 * time.Minute)),
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodRS512, claims)
+	token.Header["kid"] = c.key.Credentials.Kid
+	tokenString, err := token.SignedString(c.privateKey)
+	if err != nil {
+		return "", err
+	}
+	return tokenString, nil
+}
+
+// requestToken makes a request to the SA token API
+func (c *KeyFlow) requestToken(grant, assertion string) (*http.Response, error) {
+	body := url.Values{}
+	body.Set("grant_type", grant)
+	body.Set("assertion", assertion)
+	payload := strings.NewReader(body.Encode())
+	req, err := http.NewRequest(http.MethodPost, c.config.TokenUrl, payload)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	return c.doer(&http.Client{}, req, c.config.ClientRetry)
+}
+
+// parseTokenResponse parses the response from the server
+func (c *KeyFlow) parseTokenResponse(res *http.Response) error {
+	if res == nil {
+		return fmt.Errorf("received bad response from API")
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("received: %+v", res)
+	}
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	c.token = &TokenResponseBody{}
+	return json.Unmarshal(body, c.token)
+}
+
+// validateToken returns true if token is valid
+func (c *KeyFlow) validateToken(token string) (bool, error) {
+	if token == "" {
+		return false, nil
+	}
+	if _, err := c.parseToken(token); err != nil {
+		if strings.Contains(err.Error(), "401") {
+			c.token = new(TokenResponseBody)
```

Comment:
Why is this here?

Reply:
`parseToken`  calls `getJwksJSON` which makes a request to get jwks from the service accounts API jwks endpoint using the token. If this call returns 401 unauthorized it means that the token that is being parsed was not issued by the Service Accounts API so the client token field (containing `AccessToken` and `RefreshToken`) should be reset